Introduce `WP_List_Table::get_views_links()`.

costdev opened the pull request:
This PR introduces `WP_List_Table::get_views_links()` to abstract the link markup generation for list tables.

Includes PHPUnit tests with full line/branch coverage.

Trac ticket: https://core.trac.wordpress.org/ticket/42066

Commit:
Add `WP_List_Table::get_views_links()`.

Commit:
Users: Use `::get_views_links()` in list table.

Commit:
MS Users: Use `::get_views_links()` in list table.

Commit:
Plugins: Use `::get_views_links()` in list table.

Commit:
Plugin Install: Use `::get_views_links()` in list table.

Commit:
Theme Install: Use `::get_views_links()` in list table.

Commit:
MS Themes: Use `::get_views_links()` in list table.

Commit:
Tests: Replace `\'` with `"` to fix failures.

Commit:
MS Sites: Use `::get_views_links()` in list table.

peterwilsoncc commented on src/wp-admin/includes/class-wp-list-table.php:
```diff
@@ -1503,6 +1503,78 @@ public function ajax_response() {
+
+		$views_links = array();
+		foreach ( $link_data as $view => $link ) {
+			if ( ! isset( $link['url'] ) || ! is_string( $link['url'] ) || '' === trim( $link['url'] ) ) {
```
```suggestion
			if ( empty( $link['url'] ) || ! is_string( $link['url'] ) || '' === trim( $link['url'] ) ) {
```

:1234: Applies to label too

peterwilsoncc commented on src/wp-admin/includes/class-wp-list-table.php:
```diff
@@ -1503,6 +1503,78 @@ public function ajax_response() {
+						/* translators: %1$s: The argument name. %2$s: The view name. */
+						__( 'The <code>%1$s</code> argument must be a non-empty string for <code>%2$s</code>.' ),
+						'url',
+						$view
```
```suggestion
						esc_html( $view )
```

:1234: Applies below too.

peterwilsoncc commented on src/wp-admin/includes/class-wp-list-table.php:
```diff
@@ -1503,6 +1503,78 @@ public function ajax_response() {
+
+			$views_links[ $view ] = sprintf(
+				'<a href="%s"%s>%s</a>',
+				$link['url'],
```
```suggestion
				esc_url( $link['url'] ),
```

I like to set up CPTs with the name `"` or `'` to test for things like this. I am joking... probably.

In the core classes you've got them escaped going in to this function but there's no guarantee of that from extenders so it's safer to escape here too. It effectively runs `esc_url( esc_url() )` but `esc_url()` can deal with that pretty happily.

peterwilsoncc commented on tests/phpunit/tests/admin/wpListTable.php:
```diff
@@ -0,0 +1,226 @@
+						'label'   => 'All',
+						'current' => true,
+					),
+					'activated' => array(
+						'url'   => add_query_arg( 'status', 'activated', 'https://example.org/' ),
+						'label' => 'Activated',
+					),
+				),
+				'expected'  => array(
+					'all'       => '<a href="https://example.org/" class="current" aria-current="page">All</a>',
+					'activated' => '<a href="https://example.org/?status=activated">Activated</a>',
+				),
```
```suggestion
					'activated' => array(
						'url'   => add_query_arg( 'status', 'activated', 'https://example.org/' ),
						'label' => 'Activated',
					),
					'my activated' => array(
						'url'   => add_query_arg( array( 'author' => 1, 'status' => 'activated' ), 'https://example.org/' ),
						'label' => 'My activated',
					),
				),
				'expected'  => array(
					'all'       => '<a href="https://example.org/" class="current" aria-current="page">All</a>',
					'activated' => '<a href="https://example.org/?status=activated">Activated</a>',
					'my activated' => '<a href="https://example.org/?status=activated&amp;author=1">Activated</a>',
				),
```

Something to ensure it's escaped for output -- I think `esc_url()` escapes to `&amp;` but it would be good to check.

peterwilsoncc commented on src/wp-admin/includes/class-wp-list-table.php:
```diff
@@ -1503,6 +1503,78 @@ public function ajax_response() {
+	 * }
+	 * @return array An array of link markup. Keys match the $link_data input array.
+	 */
+	protected function get_views_links( $link_data = array() ) {
```
The list of what is `public` and `protected` in this class is a little random. Was there a reason you went with `protected` here?

Actual question, not a passive change request ;)

costdev replied:
Yep - the `get_views()` methods in each of the List Tables are set to `protected`, so I just used the same for `get_views_links()`.

peterwilsoncc replied:
Thanks, that makes sense.

peterwilsoncc commented on src/wp-admin/includes/class-wp-list-table.php:
```diff
@@ -1503,6 +1503,78 @@ public function ajax_response() {
+	 * @return array An array of link markup. Keys match the $link_data input array.
+	 */
+	protected function get_views_links( $link_data = array() ) {
+		if ( ! $link_data || ! is_array( $link_data ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				sprintf(
+					/* translators: %s: The $link_data argument. */
+					__( 'The <code>%s</code> argument must be a non-empty array.' ),
+					'$link_data'
+				),
+				'6.1.0'
+			);
+
+			return array( '' );
+		}
```
```suggestion
		if ( ! is_array( $link_data ) ) {
			_doing_it_wrong(
				__METHOD__,
				sprintf(
					/* translators: %s: The $link_data argument. */
					__( 'The <code>%s</code> argument must be an array.' ),
					'$link_data'
				),
				'6.1.0'
			);

			return array( '' );
		}
```

My logic being that if a plugin conditionally adds links based on options, enabled modules, the type of user account, etc then reaching this point without any links may be perfectly reasonable.

peterwilsoncc reviewed (commented):
Added a few notes inline.

A bit of extra escaping is needed as the highest priority item.

Commit:
WP_List_Table: Allow empty `$link_data` array.

Commit:
WP_List_Table: Use `empty()` for URL and label.

Commit:
WP_List_Table: Improve escaping.

peterwilsoncc reviewed (commented):
Thanks for the follow up.

My one last thought is about tests. There don't seem to be any tests for `get_views()` in the core test suite. 

Would it be possible to add some to ensure the markup is equivalent on the old vs new?

Commit:
Add tests for `WP_List_Table::get_views_links()`.

Commit:
Add tests for `WP_MS_Sites_List_Table::get_views()`.

Commit:
Add tests for `WP_Users_List_Table::get_views()`.

Commit:
Add tests for `WP_MS_Users_List_Table::get_views()`.

Commit:
Add tests for `WP_Plugins_List_Table::get_views()`.

Commit:
Add tests for `WP_Plugin_Install_List_Table::get_views()`.

Commit:
Add tests for `WP_Theme_Install_List_Table::get_views()`.

Commit:
Add tests for `WP_MS_Themes_List_Table::get_views()`.

Commit:
Add tests for `WP_Comments_List_Table::get_views()`.

Commit:
Comments: Use `::get_views_links()` in list table.

Commit:
Add tests for `WP_Post_Comments_List_Table::get_views()`.

Commit:
Add tests for `WP_Posts_List_Table::get_views()`.

Commit:
Posts: Use `::get_views_links()` in list table.

`WP_Posts_List_Table::get_edit_link()` is still used by:
- `::column_author()`
- `::column_default()`

These can be replaced to use `get_views_links()`, but this would lose
context.

This is another use case for a general function like the proposed
`wp_create_links()`.

Commit:
Add tests for `WP_Privacy_Requests_List_Table::get_views()`.

Commit:
Privacy Requests: Use `::get_views_links()` in list table.

Commit:
Tests: Replace `&` with `&#038;` to fix failures.

Commit:
Tests: Fix docblock alignment.

costdev commented on tests/phpunit/tests/admin/wpListTable.php:
```diff
@@ -104,4 +122,226 @@ public function data_should_only_add_primary_column_when_needed() {
 		return $datasets;
 	}
+
+	/**
```
This alignment is fixed in a commit which I'll push along with commits to fix the failing tests.

costdev replied:
Fixed in [92a05b4](https://github.com/WordPress/wordpress-develop/pull/2828/commits/92a05b4a83aeb81644e7b2e0b6eb29a709e76561).

costdev commented on src/wp-admin/includes/class-wp-posts-list-table.php:
```diff
@@ -331,16 +331,16 @@ protected function get_views() {
 				number_format_i18n( $this->user_posts_count )
 			);
 
-			$mine = $this->get_edit_link( $mine_args, $mine_inner_html, $class );
```
FYI: `WP_Posts_List_Table::get_edit_link()` is still used by:
- `WP_Posts_List_Table::column_author()`
- `WP_Posts_List_Table::column_default()`

These can be replaced to use `get_views_links()`, but this would lose context.

This is another use case for a general function like the proposed [wp_create_links()](https://github.com/WordPress/wordpress-develop/pull/2380).

westonruter replied:
With the removal of `$class` being passed in here, it seems the variable is no longer used. This is was flagged by PHPStan. In https://github.com/WordPress/wordpress-develop/pull/11151 I've included https://github.com/WordPress/wordpress-develop/pull/11151/changes/cb7a8c3c0308e052848ffc97a5e641da12721b04 to address this.

costdev commented on tests/phpunit/tests/multisite/wpMsUsersListTable.php:
```diff
@@ -0,0 +1,106 @@
+		 */
+		public function test_get_views_should_return_views_by_default() {
+			$expected = array(
+				'all'   => '<a href="http://example.org/wp-admin/network/users.php" class="current" aria-current="page">All <span class="count">(1)</span></a>',
+				'super' => '<a href="http://example.org/wp-admin/network/users.php?role=super">Super Admin <span class="count">(1)</span></a>',
```
These counts are based on a local install. The test is faulty and just needs to use `get_user_count()` and `count( get_super_admin() )` instead. I'll push this fix along with commits to fix docblock alignment and the other failing tests.

costdev replied:
Fixed in [be0780b](https://github.com/WordPress/wordpress-develop/pull/2828/commits/be0780b54a7325231e8baea27de462fac6ecb213).

Commit:
Tests: Get actual user counts in test data.

This test was previously faulty and used the counts from a local install.

This has now been resolved using:
- `get_user_count()` for 'all'
- `count( get_super_admins() )` for 'super'

peterwilsoncc commented on tests/phpunit/tests/multisite/wpMsUsersListTable.php:
```diff
@@ -0,0 +1,109 @@
+<?php
+
+if ( is_multisite() ) :
```
`@group ms-required` should prevent this from running in single site mode. 


:1234: Applies elsewhere too.

costdev replied:
Good point! Fixed in [d26fdc0](https://github.com/WordPress/wordpress-develop/pull/2828/commits/d26fdc07e5bcb0311b69b463269f73cb56bb58e8).

peterwilsoncc reviewed (commented):
LGTM but a minor change is needed to prevent the tests running in single site mode using the correct group.

Tests well when I pulled down the branch and applied it. I couldn't see any visual changes switching between this branch and trunk.

Commit:
Tests: Replace `is_multisite()` with `ms-required` group.

dream-encode commented:
Merged into core in https://core.trac.wordpress.org/changeset/54215.
